src/Tes.Runner/Executor.cs

```diff
@@ -56,14 +56,28 @@ private NodeTask DeserializeNodeTask(string nodeTask)
             }
         }
 
-        public async Task<long> UploadOutputsAsync()
+        private async Task<long> TransferIOCommonAsync(bool skipMissingSources, string? metricsFormat, TransferIOAsyncDelegate asyncDelegate)
         {
             var memoryBufferChannel = await MemoryBufferPoolFactory.CreateMemoryBufferPoolAsync(blobPipelineOptions.MemoryBufferCapacity, blobPipelineOptions.BlockSizeBytes);
 
-            return await UploadOutputsAsync(memoryBufferChannel);
+            var bytesTransfered = await asyncDelegate(memoryBufferChannel, skipMissingSources);
+
+            if (!string.IsNullOrWhiteSpace(tesNodeTask.MetricsFilename) && !string.IsNullOrWhiteSpace(metricsFormat))
+            {
+                await new MetricsFormatter(tesNodeTask.MetricsFilename, metricsFormat).Write(bytesTransfered);
+            }
+
+            return bytesTransfered;
+        }
+
+        private delegate ValueTask<long> TransferIOAsyncDelegate(Channel<byte[]> memoryBufferChannel, bool skipMissingSources);
+
+        public Task<long> UploadOutputsAsync(bool skipMissingSources)
+        {
+            return TransferIOCommonAsync(skipMissingSources, tesNodeTask.OutputsMetricsFormat, UploadOutputsAsync);
```

Comment:
I am not fully convinced this method is a good abstraction.  At this level, it makes sense to me to keep the  upload and download calls implemented explicitly. Although now they look very similar, they may diverge in the future with other parameters and special scenarios. This implementation makes it harder to maintain as now the call is delegate. I would however create a method to handle the formatting of the metrics, which in reality the only additional thing this method is doing.